### PR TITLE
Fix system prompt with added tools

### PR DESCRIPTION
**Pull request by enyst**
### Before:

When adding a tool at conversation restore:
- `base_state.json` is **updated** with the new tool
- `SystemPromptEvent` json is **not updated**
- tools sent to the LLM **contain** the new tool spec

image

### After:

[Persistence Probe Results](https://github.com/OpenHands/software-agent-sdk/blob/d25a022a25a9a266eda71e3aebc03815bcc48d4e/.pr/artifacts/20260121-223755/README.md)

### References:

- #1451
- #1778

---

## Proposed Solution

Introduce `SystemPromptUpdateEvent` — an append-only event that records system prompt or tool changes on conversation restore.

### New Event

- `SystemPromptUpdateEvent` mirrors `SystemPromptEvent` structure:
  - `source`: `"agent"`
  - `system_prompt`: `TextContent`
  - `tools`: `list[ToolDefinition]`
  - `reason`: `SystemPromptUpdateReason` enum (required)

- `SystemPromptUpdateReason` enum values:
  - `TOOLS_CHANGED`

### Behavior

- On conversation restore, `Agent.init_state()` compares runtime config vs persisted `SystemPromptEvent`
- If tools or system prompt differ → emit `SystemPromptUpdateEvent` with appropriate reason
- If unchanged → no event emitted (idempotent)

### Message Conversion

- `events_to_messages()` finds the **latest** system prompt event (original or update)
- Only the latest is converted to an LLM message; earlier ones are skipped
- System message always appears first in the message list

### Error Handling

- If events exist but no `SystemPromptEvent` found → `RuntimeError` (corrupted state)

### Exports

- `SystemPromptUpdateEvent` and `SystemPromptUpdateReason` exported from `openhands.sdk.event`

## Checklist

- [x] If the PR is changing/adding functionality, are there tests to reflect this?
- [x] If there is an example, have you run the example to make sure that it works?
- [ ] If there are instructions on how to run the code, have you followed the instructions and made sure that it works?
- [ ] If the feature is significant enough to require documentation, is there a PR open on the OpenHands/docs repository with the same branch name?
- [ ] Is the github CI passing?

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:40b0646-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-40b0646-python \
  ghcr.io/openhands/agent-server:40b0646-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:40b0646-golang-amd64
ghcr.io/openhands/agent-server:40b0646-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:40b0646-golang-arm64
ghcr.io/openhands/agent-server:40b0646-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:40b0646-java-amd64
ghcr.io/openhands/agent-server:40b0646-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:40b0646-java-arm64
ghcr.io/openhands/agent-server:40b0646-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:40b0646-python-amd64
ghcr.io/openhands/agent-server:40b0646-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-amd64
ghcr.io/openhands/agent-server:40b0646-python-arm64
ghcr.io/openhands/agent-server:40b0646-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-arm64
ghcr.io/openhands/agent-server:40b0646-golang
ghcr.io/openhands/agent-server:40b0646-java
ghcr.io/openhands/agent-server:40b0646-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `40b0646-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `40b0646-python-amd64`) are also available if needed

### CODE REVIEW

**Comment by enyst** [comment]
@openhands Make sure that for you to understand this and its conceptual background, you read both the PR description AND the references referred in the PR description, in full (all their comments too). Understand the PR. (the artefacts from testing, in .pr, will be automatically removed if the PR is approved)

Do a /codereview-roasted on it.

**Comment by enyst** [comment]
## 🟡 **Taste Rating: Acceptable**

The core design is sound — append-only events with "latest wins" semantics is the right approach for an event log. This solves a real problem (persisted `SystemPromptEvent.tools` becoming stale after restore-time tool changes) with a principled solution that maintains the invariant that the event log is the source of truth.

---

### **[GOOD TASTE]** — What's done well

1. **Append-only design** — Instead of mutating existing events (which would be a disaster for an event log), you append a new event. This is correct.

2. **"Latest wins" semantics** — Simple, clear rule: the latest system prompt event determines what the LLM sees. No complex merging logic. Easy to reason about.

3. **Idempotent behavior** — If tools haven't changed, no event is emitted. Phase C in the probe demonstrates this correctly.

4. **The probe script** — Actually testing the behavior end-to-end with real persistence is excellent engineering practice. The Markdown report makes the behavior crystal clear.

---

### **[CRITICAL ISSUES]** — Must fix

**[agent.py, Lines 161-164] 🔧 Tool comparison is incomplete:**
```python
persisted_tool_names = {t.name for t in latest_system_event.tools}
current_tool_names = {t.name for t in current_tools}
tools_changed = persisted_tool_names != current_tool_names
```
This only compares tool *names*, not tool *definitions*. If a tool's parameters or description change (e.g., new parameter added, schema updated), no `SystemPromptUpdateEvent` is emitted. The LLM will see the new schema, but the event log won't record this change — exactly the inconsistency this PR is trying to fix.

**Suggestion:** Compare tool definitions more thoroughly, or document this as intentional behavior with a comment explaining why name-only comparison is sufficient.

---

### **[IMPROVEMENT OPPORTUNITIES]** — Should fix

**[system.py, Lines 21-70 vs 73-136] 🔧 Code duplication:**
`SystemPromptEvent` and `SystemPromptUpdateEvent` have nearly identical structure and visualization code. The `visualize` property and `__str__` method are copy-pasted with minor changes. This violates DRY.

**Suggestion:** Extract a base class or mixin for the shared visualization logic, or at minimum extract the tool visualization loop into a helper function.

---

**[system.py, Lines 13-18] 🔧 Enum design:**
```python
class SystemPromptUpdateReason(str, Enum):
    TOOLS_CHANGED = "tools_changed"
    SYSTEM_PROMPT_CHANGED = "system_prompt_changed"
    TOOLS_AND_SYSTEM_PROMPT_CHANGED = "tools_and_system_prompt_changed"
```
This creates a combinatorial explosion if you add more reasons. Consider using a `frozenset` of flags or separate boolean fields instead.

---

**[base.py, Lines 105-109 vs agent.py Lines 144-147] 🔧 Inconsistent iteration direction:**
```python
# base.py - forward iteration
for event in events:
    if isinstance(event, (SystemPromptEvent, SystemPromptUpdateEvent)):
        latest_system_event = event

# agent.py - backward iteration  
for event in reversed(events):
    if isinstance(event, (SystemPromptEvent, SystemPromptUpdateEvent)):
        latest_system_event = event
        break
```
The backward iteration with `break` is more efficient (O(1) for finding the latest in typical cases). The forward iteration is O(n) always. Pick one approach and use it consistently.

---

### **[STYLE NOTES]** — Minor

**[tests/cross/test_agent_loading.py] 🔍 Missing test coverage:**
The tests only cover tool changes. There's no test for when the system prompt *text* changes (e.g., different `system_prompt_kwargs`). The code handles `SYSTEM_PROMPT_CHANGED` but it's untested.

---

### **VERDICT:**
✅ **Worth merging** with the tool comparison issue addressed. The core logic is sound, and this solves a real problem that was causing confusion for trace inspection and components like `APIBasedCritic`.

### **KEY INSIGHT:**
The fundamental insight — that system prompt changes on restore should be *recorded* in the event log rather than silently applied — is correct and aligns with the principle that the event log should be the source of truth for what the LLM actually saw.

**Comment by enyst** [comment]
## 🟢 **Taste Rating: Good Taste**

The PR author addressed the previous feedback well. This is now a clean, focused solution.

---

### **[GOOD TASTE]** — What's done well

1. **Append-only design** — Instead of mutating existing events, you append a new `SystemPromptUpdateEvent`. This is correct for an event log.

2. **"Latest wins" semantics** — Simple, clear rule: the latest system prompt event determines what the LLM sees. No complex merging logic.

3. **Idempotent behavior** — If tools haven't changed, no event is emitted. Phase C in the probe demonstrates this correctly.

4. **Consistent iteration** — Now uses backward iteration with `break` in both `_maybe_emit_system_prompt_update()` and `events_to_messages()`. ✅ Fixed.

5. **Simplified enum** — Single value `TOOLS_CHANGED` keeps scope focused. ✅ Fixed.

6. **Documented the name-only comparison** — The NOTE explains that schema changes within the same tool name won't trigger an update, covering the common add/remove case. ✅ Fixed.

7. **The probe script** — End-to-end testing with real persistence and a Markdown report is excellent engineering practice.

---

### **[ACCEPTABLE DESIGN DECISIONS]**

**Tool comparison by name only** — This is fine. The common case is adding/removing tools on restore. If a tool's parameters change within the same name, that's a schema evolution that doesn't need to be tracked as a "tool change" event. The LLM will still see the updated schema; we just don't emit a noisy update event for it.

**RuntimeError for missing SystemPromptEvent** — This is a reasonable guard. Any conversation created by this SDK will have a `SystemPromptEvent` as the first event. If it's missing, the state is genuinely corrupted.

---

### **[MINOR STYLE NOTES]**

**[system.py, Lines 66-134] Code duplication:**
`SystemPromptEvent` and `SystemPromptUpdateEvent` have similar `visualize` and `__str__` methods. Not a blocker — the duplication is contained and the classes may diverge over time.

---

### **VERDICT:**
✅ **Ready to merge.** The core design is sound, the implementation is clean, and the feedback has been addressed. This solves a real problem (persisted `SystemPromptEvent.tools` becoming stale after restore-time tool changes) with a principled, append-only solution.

### **KEY INSIGHT:**
The fundamental insight — that tool changes on restore should be *recorded* in the event log rather than silently applied — is correct and maintains the invariant that the event log is the source of truth for what the LLM actually saw.

File: openhands-sdk/openhands/sdk/agent/agent.py [hunk 1]
```diff
@@ -120,6 +121,64 @@ def init_state(
...
+
+        # Compare tool names with persisted values
+        # NOTE: We compare names only, not full definitions. Schema changes within
+        # the same tool name won't trigger an update. This covers the common case
+        # of adding/removing tools on restore.
+        persisted_tool_names = {t.name for t in latest_system_event.tools}
+        current_tool_names = {t.name for t in current_tools}
+
+        if persisted_tool_names == current_tool_names:
+            return  # Tools unchanged, no update needed
```
[the remark this answers is not in the document]
Comment (enyst) [reply openhands-sdk/openhands/sdk/agent/agent.py:167]: This is the status quo. It's likely the most common case, and it's the case of end user use.

File: openhands-sdk/openhands/sdk/agent/agent.py [hunk 2]
```diff
@@ -120,6 +121,65 @@ def init_state(
...
+
+        logger.info(
```
Comment (enyst) [inline openhands-sdk/openhands/sdk/agent/agent.py]:
```suggestion
        logger.debug(
```

File: tests/cross/test_agent_loading.py [hunk 3]
```diff
@@ -641,3 +641,113 @@ def test_conversation_restart_with_different_agent_context():
             "You current working directory is: /Users/jpshack"
             in new_conversation.agent.agent_context.system_message_suffix
         )
+
+
+def test_conversation_restore_emits_system_prompt_update_when_tools_change():
+    """Test restore with different tools emits SystemPromptUpdateEvent.
+
+    This ensures the persisted event log accurately reflects what the LLM sees,
+    maintaining the invariant that the event log is the source of truth.
```
[the remark this answers is not in the document]
Comment (enyst) [reply tests/cross/test_agent_loading.py:650]: Good catch! I added three tests to cover these edge cases:

1. **Tool removal** (`test_conversation_restore_emits_update_when_tools_removed`): A+B → A emits update ✅
2. **Tool swap** (`test_conversation_restore_emits_update_when_tools_swapped`): A → B emits update ✅
3. **Tool reordering** (`test_conversation_restore_no_update_when_tools_reordered`): A,B → B,A does NOT emit update (set comparison) ✅

All 16 tests in `test_agent_loading.py` pass.

Commit: e1dea9e6

File: openhands-sdk/openhands/sdk/agent/agent.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (enyst) [reply openhands-sdk/openhands/sdk/agent/agent.py:167]: I left this unsolved, to be visible, in case you may feel differently, Xingyao?
